contracts/RelayHub.sol

```diff
@@ -58,130 +57,60 @@ contract RelayHub is IRelayHub {
 
-    mapping(address => Relay) private relays;
     mapping(address => uint256) private balances;
 
     string public version = "1.0.0";
 
-    constructor (uint256 _gtxdatanonzero) public {
+
+    EIP712Sig private eip712sig;
+    StakeManager private stakeManager;
+    constructor (uint256 _gtxdatanonzero, StakeManager _stakeManager) public {
+        eip712sig = new EIP712Sig(address(this));
+        stakeManager = _stakeManager;
         gtxdatanonzero = _gtxdatanonzero;
     }
 
     function calldatagascost() private view returns (uint256) {
         return GTRANSACTION + msg.data.length * gtxdatanonzero;
     }
 
-    function stake(address relay, uint256 unstakeDelay) external payable {
-        if (relays[relay].state == RelayState.Unknown) {
-            require(msg.sender != relay, "relay cannot stake for itself");
-            relays[relay].owner = msg.sender;
-            relays[relay].state = RelayState.Staked;
-
-        } else if ((relays[relay].state == RelayState.Staked) || (relays[relay].state == RelayState.Registered)) {
-            require(relays[relay].owner == msg.sender, "not owner");
-
-        } else {
-            revert("wrong state for stake");
-        }
-
-        // Increase the stake
-
-        uint256 addedStake = msg.value;
-        relays[relay].stake += addedStake;
-
-        // The added stake may be e.g. zero when only the unstake delay is being updated
-        require(relays[relay].stake >= MINIMUM_STAKE, "stake lower than minimum");
-
-        // Increase the unstake delay
-
-        require(unstakeDelay >= MINIMUM_UNSTAKE_DELAY, "delay lower than minimum");
-        require(unstakeDelay <= MAXIMUM_UNSTAKE_DELAY, "delay higher than maximum");
 
-        require(unstakeDelay >= relays[relay].unstakeDelay, "unstakeDelay cannot be decreased");
-        relays[relay].unstakeDelay = unstakeDelay;
-
-        emit Staked(relay, relays[relay].stake, relays[relay].unstakeDelay);
+    function registerRelayServer(uint256 baseRelayFee, uint256 pctRelayFee, string calldata url) external {
+        address relayManager = msg.sender;
+        require(
+            stakeManager.isRelayManagerStaked(relayManager, MINIMUM_STAKE, MINIMUM_UNSTAKE_DELAY),
+            "relay manager not staked"
+        );
+        require(workerCount[relayManager] > 0, "no relay workers");
+        emit RelayServerRegistered(relayManager, baseRelayFee, pctRelayFee, url);
     }
 
-    function registerRelay(uint256 baseRelayFee, uint256 pctRelayFee, string memory url) public {
-        address relay = msg.sender;
+    function addRelayWorkers(address[] calldata newRelayWorkers) external {
+        address relayManager = msg.sender;
+        workerCount[relayManager] = workerCount[relayManager] + newRelayWorkers.length;
+        require(workerCount[relayManager] <= MAX_WORKER_COUNT, "too many workers");
 
-        // solhint-disable-next-line avoid-tx-origin
-        require(relay == tx.origin, "Contracts cannot register as relays");
```

Comment:
you removed tx.origin check on register, but we MUST have it for workers, so either:
- take signatures instead of addresses when registering (one-time expensive check when registering a relay - on owner's balance)
- check `msg.sender == tx.origin` when making a relayed call (each check on every call, on sender's balance)